src/service/resource.js

```diff
@@ -560,6 +566,32 @@ export class Resource {
   }
 
   /**
+   * @return {!Promise} resolves when underlying element is built and within the
+   *    range specified by implementation's renderOutsideViewport
+   */
+  whenWithinRenderOutsideViewport() {
+    if (this.renderOutsideViewportPromise_) {
+      return this.renderOutsideViewportPromise_;
+    }
+    return this.renderOutsideViewportPromise_ = new Promise(resolver => {
```

Comment:
There's still a change for a never resolving promise:

1. Element is already laid out
2. Whatever calls `#whenWithinRenderOutsideViewport`

Since we're laid out, we'll never call `#renderOutsideViewport` again, and the promise will never resolve. We need to check state and resolve immediately if laid out.